### PR TITLE
GEODE-10162: Try to find and set version tag if needed

### PR DESCRIPTION
**Pull request by pivotal-eshu**
* In geode, version tags are not sent to receiver during GII. When an
  operation is retried on the receiver, a valid tag is needed so that the
  operation can be distributed to other peers. Valid version tage will
  be searched in the cluster and set in the event to make sure the
  operation is distributed.

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java [hunk 1]
```diff
@@ -526,21 +526,36 @@ public boolean hasSeenEvent(EntryEventImpl event) {
     isDuplicate = getEventTracker().hasSeenEvent(event);
     if (isDuplicate) {
       markEventAsDuplicate(event);
+      if (event.getVersionTag() == null && event.getRegion().getConcurrencyChecksEnabled()
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java:529]: We could move all this checks to findAndSetVersionTag()
Then, it will be:

public boolean hasSeenEvent(EntryEventImpl event) {
    boolean isDuplicate;
    isDuplicate = getEventTracker().hasSeenEvent(event);
    if (isDuplicate) {
      markEventAsDuplicate(event);
    } 
    if (event.isPossibleDuplicate()) {
      findAndSetVersionTag(event);
    }
    return isDuplicate;
  }

  void findAndSetVersionTag(EntryEventImpl event) {
    if (event.isPossibleDuplicate() && event.getRegion().getConcurrencyChecksEnabled()
        && event.getVersionTag() == null && event.getEventId() != null) {
      boolean isBulkOp = event.getOperation().isPutAll() || event.getOperation().isRemoveAll();
      VersionTag tag =
          FindVersionTagOperation.findVersionTag(event.getRegion(), event.getEventId(), isBulkOp);
      if (tag == null) {
        logger.info("No version tag can be found in cluster when retrying the following event {}",
            event);
      }
      event.setVersionTag(tag);
    }
  }

**Review by gesterzhou — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java [hunk 2]
```diff
@@ -526,21 +526,36 @@ public boolean hasSeenEvent(EntryEventImpl event) {
...
+          && event.getEventId() != null) {
+        // For a retry, versionTag can be missing possibly due to no versionTag recorded during gii.
+        findAndSetVersionTag(event);
+        if (event.getVersionTag() == null) {
+          logger.info("No version tag can be found in cluster when retrying the following event {}",
+              event);
```
Comment (DonalEvans) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java:535]: Is this likely to be logged often? If there are situations where it could be logged a lot, it might be better to make this debug level.

File: geode-core/src/test/java/org/apache/geode/internal/cache/DistributedRegionTest.java [hunk 3]
```diff
@@ -260,4 +266,132 @@ public void validateAsynchronousEventDispatcherShouldThrowExceptionWhenDispatche
         .hasMessage("Parallel Gateway Sender " + senderId
             + " can not be used with replicated region " + regionPath);
   }
+
+  @Test
+  public void hasSeenEventDoseNotFindAndSetVersionTagIfFoundInEventTrackerAndVersionTagIsSet() {
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/DistributedRegionTest.java:271]: Typo here and in other tests added, should be "DoesNot" rather than "DoseNot"

**Review by DonalEvans — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java (hunk 2 above)
Comment (pivotal-eshu) [reply geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java:535]: No, we should not see this -- only a very corner case.

File: geode-core/src/test/java/org/apache/geode/internal/cache/DistributedRegionTest.java (hunk 3 above)
Comment (pivotal-eshu) [reply geode-core/src/test/java/org/apache/geode/internal/cache/DistributedRegionTest.java]: Good catch!

**Review by kirklund — APPROVED** [review]
